### DISCUSSION TITLE
如何在索尼 PS 平台上使用 Puerts？

### DISCUSSION THREAD

**Discussion by shpz** [Tencent/puerts#1193] [Q&A] [answered]
公司项目现在在快速迭代期，考虑接入 Puerts。
但目前有个问题，PS 上怎么搞定 v8/quickjs/nodejs。
希望社区大佬不吝赐教，感谢。

**Comment by chexiongsheng** [comment, marked as answer]
如果是quickjs，很简单，即使些纯c代码，跨平台兼容没啥问题（ps能支持c语言写代码，就可以支持）。这可以作为一个保底。

如果是v8，首先它本身的编译复杂，不仅仅是c++代码，它还需要编译些工具来辅助，比如它里头有些代码是用torque语言（应该就只有v8自己用）写，需要先用本地编译器编译出torque编译器，然后用torque编译器把torque语言编译成c++再编译。

其次它本身的jit的cpu、平台相关性比较大，未知支不支持的情况下，关掉最稳妥。

我看它的代码，只对window特殊处理，其它posix系统（mac，linux，android，ios）都统一，我觉得ps大体按posix也问题不大（ps系统好像是基于linux的）。

具体有两个方向。

它本身是用gn工具生成ninja配置来编译，感觉一个可行的路子是用gn选一个平台（直接编译选linux，交叉编译选android）生成ninja配置，然后修改gn配置里头的编译器，链接器等的指向。当然，如果你对gn很熟悉，直接写个ps5的支持也行。

另外一个可行的路子是这个项目：https://github.com/bnoordhuis/v8-cmake ，看维护人员有nodejs的人。
这个是v8的cmake编译支持。而cmake是能支持ps的。
要注意这个仓库并不是每个版本都是好的，我试过一些版本在linux编译就失败。亲测这个版本至少linux是可用的：
https://github.com/bnoordhuis/v8-cmake/commit/8dfb23f20c466f290d454c4cd63ef93baa49765c

**Reply by shpz** [reply]
> 如果是quickjs，很简单，即使些纯c代码，跨平台兼容没啥问题（ps能支持c语言写代码，就可以支持）。这可以作为一个保底。
> 
> 如果是v8，首先它本身的编译复杂，不仅仅是c++代码，它还需要编译些工具来辅助，比如它里头有些代码是用torque语言（应该就只有v8自己用）写，需要先用本地编译器编译出torque编译器，然后用torque编译器把torque语言编译成c++再编译。
> 
> 其次它本身的jit的cpu、平台相关性比较大，未知支不支持的情况下，关掉最稳妥。
> 
> 我看它的代码，只对window特殊处理，其它posix系统（mac，linux，android，ios）都统一，我觉得ps大体按posix也问题不大（ps系统好像是基于linux的）。
> 
> 具体有两个方向。
> 
> 它本身是用gn工具生成ninja配置来编译，感觉一个可行的路子是用gn选一个平台（直接编译选linux，交叉编译选android）生成ninja配置，然后修改gn配置里头的编译器，链接器等的指向。当然，如果你对gn很熟悉，直接写个ps5的支持也行。
> 
> 另外一个可行的路子是这个项目：https://github.com/bnoordhuis/v8-cmake ，看维护人员有nodejs的人。 这个是v8的cmake编译支持。而cmake是能支持ps的。 要注意这个仓库并不是每个版本都是好的，我试过一些版本在linux编译就失败。亲测这个版本至少linux是可用的： [bnoordhuis/v8-cmake@8dfb23f](https://github.com/bnoordhuis/v8-cmake/commit/8dfb23f20c466f290d454c4cd63ef93baa49765c)

感谢大佬的回复！
我更倾向使用 v8，这样我在发包的时候不用考虑切换 js 后端和一些其他问题。
我对 gn 没有了解，对 cmake 有一定了解，我会从 cmake 的方案开始尝试。
后续如果有进展会分享出来。

**Comment by chexiongsheng** [comment]
有什么进展呢？
